File: Application/Services/ComplaintService.cs
using Core.DTOs.Complaint;
using Core.DTOs.ITop;
using Core.Entities.Complaint;
using Core.Enums;
using Core.Interfaces.Repositories;
using Core.Interfaces.Services;
using Microsoft.AspNetCore.Http;

namespace Application.Services
{
    public class ComplaintService : IComplaintService
    {
        private readonly IComplaintRepository _complaintRepo;
        private readonly ILocationRepository _locationRepo;
        private readonly IInternalUserRepository _userRepo;
        private readonly INotificationRepository _notificationRepo;
        private readonly IEscalationService _escalationService;
        private readonly IStorageService _storage;
        private readonly IITopTicketAdapter _itopAdapter;
        private readonly IPushNotificationService _pushNotificationService;

        public ComplaintService(
            IComplaintRepository complaintRepo,
            ILocationRepository locationRepo,
            IInternalUserRepository userRepo,
            INotificationRepository notificationRepo,
            IEscalationService escalationService,
            IStorageService storage,
            IITopTicketAdapter itopAdapter,
            IPushNotificationService pushNotificationService)
        {
            _complaintRepo = complaintRepo;
            _locationRepo = locationRepo;
            _userRepo = userRepo;
            _notificationRepo = notificationRepo;
            _escalationService = escalationService;
            _storage = storage;
            _itopAdapter = itopAdapter;
            _pushNotificationService = pushNotificationService;
        }

        public async Task<ComplaintResponseDto> SubmitComplaintAsync(Guid citizenId, SubmitComplaintDto dto)
        {
            var block = await _locationRepo.GetBlockByIdAsync(dto.BlockId)
                ?? throw new KeyNotFoundException("Block not found.");

            var refNumber = await _complaintRepo.GenerateRefNumberAsync();

            var complaint = new Complaint
            {
                Id = Guid.NewGuid(),
                RefNumber = refNumber,
                Title = dto.Title,
                Description = dto.Description,
                CategoryId = dto.CategoryId,
                BlockId = dto.BlockId,
                CitizenId = citizenId,
                Priority = dto.Priority,
                Status = ComplaintStatus.Submitted,
                SubmittedAt = DateTime.UtcNow,
                CreatedAt = DateTime.UtcNow,
                UpdatedAt = DateTime.UtcNow
            };

            await _complaintRepo.CreateAsync(complaint);

            // Reload with includes
            var created = await _complaintRepo.GetByIdAsync(complaint.Id)
                ?? throw new Exception("Failed to retrieve created complaint.");

            await CreateITopTicketMappingAsync(created);
            await NotifyAssignersAsync(created);

            return MapToDto(created);
        }

        public async Task<ComplaintResponseDto> GetByIdAsync(Guid id)
        {
            var complaint = await _complaintRepo.GetByIdAsync(id)
                ?? throw new KeyNotFoundException("Complaint not found.");
            return MapToDto(complaint);
        }

        public async Task<IEnumerable<ComplaintResponseDto>> GetMyComplaintsAsync(Guid citizenId)
        {
            var complaints = await _complaintRepo.GetByCitizenIdAsync(citizenId);
            return complaints.Select(MapToDto);
        }

        public async Task<PagedComplaintsDto> GetAllComplaintsAsync(ComplaintStatus? status, int? departmentId, int? blockId, int page, int pageSize)
        {
            var items = await _complaintRepo.GetAllAsync(
                status, departmentId, blockId, page, pageSize);
            var total = await _complaintRepo.GetTotalCountAsync(
                status, departmentId, blockId);

            return new PagedComplaintsDto
            {
                Items = items.Select(MapToDto),
                TotalCount = total,
                Page = page,
                PageSize = pageSize
            };
        }

        public async Task<ComplaintResponseDto> AssignDepartmentAsync(Guid complaintId, AssignComplaintDto dto, Guid assignedById)
        {
            var complaint = await _complaintRepo.GetByIdAsync(complaintId)
                ?? throw new KeyNotFoundException("Complaint not found.");

            if (complaint.Status != ComplaintStatus.Submitted)
                throw new InvalidOperationException("Only submitted complaints can be assigned.");

            complaint.AssignedDepartmentId = dto.DepartmentId;
            complaint.Status = ComplaintStatus.Assigned;
            complaint.AssignedAt = DateTime.UtcNow;
            complaint.UpdatedAt = DateTime.UtcNow;

            await _complaintRepo.UpdateAsync(complaint);

            await SyncStatusToITopAsync(
                complaint,
                ComplaintStatus.Assigned,
                $"Assigned to department ID {dto.DepartmentId}");
            await NotifyDepartmentAssignedAsync(complaint);

            var updated = await _complaintRepo.GetByIdAsync(complaintId)!;
            return MapToDto(updated!);
        }

        public async Task<ComplaintResponseDto> AssignAgentAsync(Guid complaintId, Guid agentId, Guid assignedById)
        {
            var complaint = await _complaintRepo.GetByIdAsync(complaintId)
                ?? throw new KeyNotFoundException("Complaint not found.");

            var agent = await _userRepo.GetByIdAsync(agentId)
                ?? throw new KeyNotFoundException("Agent not found.");

            complaint.AssignedAgentId = agentId;
            if (complaint.Status == ComplaintStatus.Assigned)
                complaint.Status = ComplaintStatus.InProgress;
            complaint.UpdatedAt = DateTime.UtcNow;

            await _complaintRepo.UpdateAsync(complaint);

            if (complaint.Status == ComplaintStatus.InProgress)
            {
                await SyncStatusToITopAsync(
                    complaint,
                    ComplaintStatus.InProgress,
                    $"Field agent assigned: {agent.FullName}");
                await NotifyCitizenAsync(
                    complaint,
                    NotificationType.ComplaintInProgress,
                    $"Your complaint {complaint.RefNumber} is now in progress.");
            }
            await NotifyUserAsync(
                SenderType.Agent,
                agent.Id,
                complaint.Id,
                NotificationType.ComplaintAssigned,
                $"Complaint {complaint.RefNumber} has been assigned to you.");

            var updated = await _complaintRepo.GetByIdAsync(complaintId)!;
            return MapToDto(updated!);
        }

        public async Task<ComplaintResponseDto> UpdateStatusAsync(
            Guid complaintId,
            UpdateComplaintStatusDto dto,
            Guid updatedById)
        {
            var complaint = await _complaintRepo.GetByIdAsync(complaintId)
                ?? throw new KeyNotFoundException("Complaint not found.");

            ValidateStatusTransition(complaint.Status, dto.Status);

            complaint.Status = dto.Status;
            complaint.UpdatedAt = DateTime.UtcNow;

            if (dto.Status == ComplaintStatus.Resolved)
                complaint.ResolvedAt = DateTime.UtcNow;
            else if (dto.Status == ComplaintStatus.Closed)
                complaint.ClosedAt = DateTime.UtcNow;

            await _complaintRepo.UpdateAsync(complaint);

            // System message
            var message = new ComplaintMessage
            {
                Id = Guid.NewGuid(),
                ComplaintId = complaintId,
                SenderType = SenderType.System,
                SenderId = updatedById,
                Message = dto.Remarks ?? $"Status updated to {dto.Status}",
                CreatedAt = DateTime.UtcNow
            };
            await _complaintRepo.AddMessageAsync(message);

            // iTop sync
            await SyncStatusToITopAsync(complaint, dto);
            await NotifyStatusChangedAsync(complaint, dto.Status);
            if (dto.Status == ComplaintStatus.Resolved || dto.Status == ComplaintStatus.Closed)
                await _escalationService.ResolveActiveForComplaintAsync(complaint.Id);

            var updated = await _complaintRepo.GetByIdAsync(complaintId)!;
            return MapToDto(updated!);
        }

        public async Task<ComplaintMediaResponseDto> UploadMediaAsync(Guid complaintId, IFormFile file, Guid uploadedById)
        {
            // Verify complaint exists
            var complaint = await _complaintRepo.GetByIdAsync(complaintId)
                ?? throw new KeyNotFoundException("Complaint not found.");

            ValidateUpload(file);

            var mediaType = DetermineMediaType(file.ContentType ?? string.Empty);
            var folder = $"complaints/{complaintId}";
            var fileName = Path.GetFileName(file.FileName);
            byte[] fileContent;

            await using (var buffer = new MemoryStream())
            {
                await file.CopyToAsync(buffer);
                fileContent = buffer.ToArray();
            }

            var filePath = await _storage.SaveFileAsync(
                new MemoryStream(fileContent),
                fileName,
                folder
            );

            var media = new ComplaintMedia
            {
                Id = Guid.NewGuid(),
                ComplaintId = complaintId,
                MediaType = mediaType,
                FileName = fileName,
                FilePath = filePath,
                FileSize = file.Length,
                MimeType = file.ContentType,
                CreatedAt = DateTime.UtcNow
            };

            // Save media directly — don't touch the complaint entity
            await _complaintRepo.AddMediaAsync(media);
            await SyncAttachmentToITopAsync(complaint, media, fileContent);

            return new ComplaintMediaResponseDto
            {
                Id = media.Id,
                MediaType = media.MediaType.ToString(),
                FileName = media.FileName,
                FileUrl = _storage.GetFileUrl(media.FilePath),
                FileSize = media.FileSize,
                CreatedAt = media.CreatedAt
            };
        }

        public async Task<ComplaintMessageResponseDto> SendMessageAsync(
            Guid complaintId,
            SendComplaintMessageDto dto,
            Guid senderId,
            bool isInternalUser)
        {
            var complaint = await _complaintRepo.GetByIdAsync(complaintId)
                ?? throw new KeyNotFoundException("Complaint not found.");

            if (string.IsNullOrWhiteSpace(dto.Message))
                throw new InvalidOperationException("Message is required.");

            var senderType = ResolveSenderType(dto.SenderType, isInternalUser);
            var message = new ComplaintMessage
            {
                Id = Guid.NewGuid(),
                ComplaintId = complaintId,
                SenderType = senderType,
                SenderId = senderId,
                Message = dto.Message.Trim(),
                IsRead = false,
                CreatedAt = DateTime.UtcNow
            };

            await _complaintRepo.AddMessageAsync(message);
            await SyncMessageToITopAsync(complaint, message);
            await NotifyMessageRecipientsAsync(complaint, message);
            return MapMessageToDto(message);
        }

        public async Task<IEnumerable<ComplaintMessageResponseDto>> GetMessagesAsync(Guid complaintId)
        {
            _ = await _complaintRepo.GetByIdAsync(complaintId)
                ?? throw new KeyNotFoundException("Complaint not found.");

            var messages = await _complaintRepo.GetMessagesAsync(complaintId);
            return messages.Select(MapMessageToDto);
        }

        public async Task<ComplaintMessageResponseDto> MarkMessageAsReadAsync(
            Guid complaintId,
            Guid messageId)
        {
            var message = await _complaintRepo.GetMessageByIdAsync(complaintId, messageId)
                ?? throw new KeyNotFoundException("Message not found.");

            message.IsRead = true;
            await _complaintRepo.UpdateMessageAsync(message);
            return MapMessageToDto(message);
        }

        public async Task<ComplaintFeedbackResponseDto> SubmitFeedbackAsync(
            Guid complaintId,
            SubmitFeedbackDto dto,
            Guid submittedById,
            bool isInternalUser)
        {
            var complaint = await _complaintRepo.GetByIdAsync(complaintId)
                ?? throw new KeyNotFoundException("Complaint not found.");

            if (complaint.Status != ComplaintStatus.Resolved &&
                complaint.Status != ComplaintStatus.Closed)
                throw new InvalidOperationException("Feedback can be submitted only after resolution.");

            if (dto.Rating < 1 || dto.Rating > 5)
                throw new InvalidOperationException("Rating must be between 1 and 5.");

            var existing = await _complaintRepo.GetFeedbackAsync(complaintId);
            if (existing != null)
                throw new InvalidOperationException("Feedback already exists for this complaint.");

            var feedback = new ComplaintFeedback
            {
                Id = Guid.NewGuid(),
                ComplaintId = complaintId,
                CitizenId = complaint.CitizenId,
                Rating = dto.Rating,
                Comments = dto.Comments,
                CollectedById = isInternalUser ? submittedById : null,
                CreatedAt = DateTime.UtcNow
            };

            await _complaintRepo.AddFeedbackAsync(feedback);
            return MapFeedbackToDto(feedback);
        }

        public async Task<ComplaintFeedbackResponseDto> GetFeedbackAsync(Guid complaintId)
        {
            _ = await _complaintRepo.GetByIdAsync(complaintId)
                ?? throw new KeyNotFoundException("Complaint not found.");

            var feedback = await _complaintRepo.GetFeedbackAsync(complaintId)
                ?? throw new KeyNotFoundException("Feedback not found.");

            return MapFeedbackToDto(feedback);
        }

        // ── Helpers ────────────────────────────────────────────────

        private MediaType DetermineMediaType(string contentType) =>
            contentType switch
            {
                var ct when ct.StartsWith("image/") => MediaType.Image,
                var ct when ct.StartsWith("video/") => MediaType.Video,
                var ct when ct.StartsWith("audio/") => MediaType.Voice,
                _ => MediaType.Document
            };

        private static void ValidateUpload(IFormFile file)
        {
            const long maxBytes = 50 * 1024 * 1024;
            if (file.Length <= 0)
                throw new InvalidOperationException("Uploaded file is empty.");
            if (file.Length > maxBytes)
                throw new InvalidOperationException("Uploaded file exceeds the 50 MB limit.");
            if (string.IsNullOrWhiteSpace(file.FileName))
                throw new InvalidOperationException("Uploaded file name is required.");
        }

        private static void ValidateStatusTransition(ComplaintStatus current, ComplaintStatus next)
        {
            var allowed = current switch
            {
                ComplaintStatus.Submitted => new[] { ComplaintStatus.Assigned, ComplaintStatus.Rejected },
                ComplaintStatus.Assigned => new[] { ComplaintStatus.InProgress, ComplaintStatus.Rejected },
                ComplaintStatus.InProgress => new[] { ComplaintStatus.Resolved, ComplaintStatus.Rejected },
                ComplaintStatus.Resolved => new[] { ComplaintStatus.Closed, ComplaintStatus.InProgress },
                _ => Array.Empty<ComplaintStatus>()
            };

            if (!allowed.Contains(next))
                throw new InvalidOperationException($"Invalid status transition from {current} to {next}.");
        }

        private static SenderType ResolveSenderType(SenderType? requested, bool isInternalUser)
        {
            if (!isInternalUser)
            {
                if (requested.HasValue && requested.Value != SenderType.Citizen)
                    throw new InvalidOperationException("Citizens can send only citizen messages.");
                return SenderType.Citizen;
            }

            if (requested == SenderType.System)
                return SenderType.System;

            return SenderType.Agent;
        }

        private async Task CreateITopTicketMappingAsync(Complaint complaint)
        {
            var result = await _itopAdapter.CreateTicketAsync(new ITopTicketCreateRequest
            {
                ComplaintId = complaint.Id,
                RefNumber = complaint.RefNumber,
                Title = complaint.Title,
                Description = complaint.Description,
                CitizenName = complaint.Citizen?.FullName ?? string.Empty,
                CitizenPhone = complaint.Citizen?.Phone ?? string.Empty,
                CategoryName = complaint.Category?.Name ?? string.Empty,
                BlockName = complaint.Block?.Name ?? string.Empty,
                Priority = complaint.Priority
            });

            if (!result.WasAttempted)
                return;

            var mapping = new ComplaintITopMapping
            {
                Id = Guid.NewGuid(),
                ComplaintId = complaint.Id,
                ITopTicketId = result.TicketId ?? string.Empty,
                ITopTicketRef = result.TicketRef ?? string.Empty,
                ITopClass = "UserRequest",
                LastSyncedAt = result.Success ? DateTime.UtcNow : null,
                SyncStatus = result.Success ? SyncStatus.Synced : SyncStatus.SyncFailed,
                LastSyncError = result.Success ? null : result.Error
            };

            await _complaintRepo.AddITopMappingAsync(mapping);
        }

        private ComplaintResponseDto MapToDto(Complaint c) => new()
        {
            Id = c.Id,
            RefNumber = c.RefNumber,
            Title = c.Title,
            Description = c.Description,
            Status = c.Status.ToString(),
            Priority = c.Priority,
            CreatedAt = c.CreatedAt,
            SubmittedAt = c.SubmittedAt,
            AssignedAt = c.AssignedAt,
            ResolvedAt = c.ResolvedAt,
            CitizenId = c.CitizenId,
            CitizenName = c.Citizen?.FullName ?? string.Empty,
            CitizenPhone = c.Citizen?.Phone ?? string.Empty,
            BlockId = c.BlockId,
            BlockName = c.Block?.Name ?? string.Empty,
            CategoryId = c.CategoryId,
            CategoryName = c.Category?.Name ?? string.Empty,
            AssignedDepartmentId = c.AssignedDepartmentId,
            AssignedDepartmentName = c.AssignedDepartment?.Name,
            AssignedAgentId = c.AssignedAgentId,
            AssignedAgentName = c.AssignedAgent?.FullName,
            Media = c.Media.Select(m => new ComplaintMediaResponseDto
            {
                Id = m.Id,
                MediaType = m.MediaType.ToString(),
                FileName = m.FileName,
                FileUrl = _storage.GetFileUrl(m.FilePath),
                FileSize = m.FileSize,
                CreatedAt = m.CreatedAt
            }).ToList()
        };

        private static ComplaintMessageResponseDto MapMessageToDto(ComplaintMessage m) => new()
        {
            Id = m.Id,
            ComplaintId = m.ComplaintId,
            SenderType = m.SenderType.ToString(),
            SenderId = m.SenderId,
            Message = m.Message,
            IsRead = m.IsRead,
            CreatedAt = m.CreatedAt
        };

        private static ComplaintFeedbackResponseDto MapFeedbackToDto(ComplaintFeedback f) => new()
        {
            Id = f.Id,
            ComplaintId = f.ComplaintId,
            CitizenId = f.CitizenId,
            Rating = f.Rating,
            Comments = f.Comments,
            CollectedById = f.CollectedById,
            CreatedAt = f.CreatedAt
        };

        private async Task SyncStatusToITopAsync(
            Complaint complaint,
            UpdateComplaintStatusDto dto)
        {
            await SyncStatusToITopAsync(complaint, dto.Status, dto.Remarks);
        }

        private async Task SyncStatusToITopAsync(
            Complaint complaint,
            ComplaintStatus newStatus,
            string? remarks)
        {
            var mapping = await _complaintRepo
                .GetITopMappingByComplaintIdAsync(complaint.Id);

            if (mapping == null || mapping.SyncStatus != SyncStatus.Synced)
                return;

            var result = await _itopAdapter.UpdateTicketAsync(new ITopTicketUpdateRequest
            {
                ITopTicketId = mapping.ITopTicketId,
                ITopClass = mapping.ITopClass,
                NewStatus = newStatus.ToString(),
                Remarks = remarks,
                ComplaintRefNumber = complaint.RefNumber
            });

            mapping.LastSyncedAt = DateTime.UtcNow;
            mapping.SyncStatus = result.Success ? SyncStatus.Synced : SyncStatus.SyncFailed;
            mapping.LastSyncError = result.Success ? null : result.Error;

            await _complaintRepo.UpdateITopMappingAsync(mapping);
        }

        private async Task SyncAttachmentToITopAsync(
            Complaint complaint,
            ComplaintMedia media,
            byte[] fileContent)
        {
            var mapping = await _complaintRepo
                .GetITopMappingByComplaintIdAsync(complaint.Id);

            if (mapping == null || mapping.SyncStatus != SyncStatus.Synced)
                return;

            var result = await _itopAdapter.CreateAttachmentAsync(new ITopAttachmentCreateRequest
            {
                ITopTicketId = mapping.ITopTicketId,
                ITopClass = mapping.ITopClass,
                FileName = media.FileName,
                MimeType = media.MimeType ?? "application/octet-stream",
                Content = fileContent,
                ComplaintRefNumber = complaint.RefNumber
            });

            if (!result.WasAttempted)
                return;

            mapping.LastSyncedAt = result.Success ? DateTime.UtcNow : mapping.LastSyncedAt;
            mapping.LastSyncError = result.Success
                ? null
                : $"Attachment sync failed for {media.FileName}: {result.Error}";

            await _complaintRepo.UpdateITopMappingAsync(mapping);
        }

        private async Task SyncMessageToITopAsync(Complaint complaint, ComplaintMessage message)
        {
            var mapping = await _complaintRepo
                .GetITopMappingByComplaintIdAsync(complaint.Id);

            if (mapping == null || mapping.SyncStatus != SyncStatus.Synced)
                return;

            var result = await _itopAdapter.AddTicketLogAsync(new ITopTicketLogRequest
            {
                ITopTicketId = mapping.ITopTicketId,
                ITopClass = mapping.ITopClass,
                ComplaintRefNumber = complaint.RefNumber,
                Message = FormatITopMessageLog(message),
                IsPrivate = message.SenderType == SenderType.System
            });

            if (!result.WasAttempted)
                return;

            mapping.LastSyncedAt = result.Success ? DateTime.UtcNow : mapping.LastSyncedAt;
            mapping.LastSyncError = result.Success
                ? null
                : $"Message sync failed for {message.Id}: {result.Error}";

            await _complaintRepo.UpdateITopMappingAsync(mapping);
        }

        private static string FormatITopMessageLog(ComplaintMessage message)
        {
            var sender = message.SenderType switch
            {
                SenderType.Citizen => "Citizen",
                SenderType.Agent => "Agent",
                SenderType.System => "System",
                _ => "Unknown"
            };

            return $"[{sender}] {message.Message}";
        }

        private async Task NotifyAssignersAsync(Complaint complaint)
        {
            var assigners = await _userRepo.GetByRoleAsync(UserRole.Assigner);
            await NotifyUsersAsync(
                assigners.Select(u => u.Id),
                SenderType.Agent,
                complaint.Id,
                NotificationType.ComplaintSubmitted,
                $"New complaint {complaint.RefNumber} has been submitted.");
        }

        private async Task NotifyDepartmentAssignedAsync(Complaint complaint)
        {
            await NotifyCitizenAsync(
                complaint,
                NotificationType.ComplaintAssigned,
                $"Your complaint {complaint.RefNumber} has been assigned to a department.");

            if (complaint.AssignedDepartmentId.HasValue)
            {
                var agents = await _userRepo.GetByRoleAndDepartmentAsync(
                    UserRole.FieldAgent,
                    complaint.AssignedDepartmentId.Value);

                await NotifyUsersAsync(
                    agents.Select(a => a.Id),
                    SenderType.Agent,
                    complaint.Id,
                    NotificationType.ComplaintAssigned,
                    $"Complaint {complaint.RefNumber} has been assigned to your department.");
            }
        }

        private async Task NotifyStatusChangedAsync(Complaint complaint, ComplaintStatus status)
        {
            if (status == ComplaintStatus.InProgress)
            {
                await NotifyCitizenAsync(
                    complaint,
                    NotificationType.ComplaintInProgress,
                    $"Your complaint {complaint.RefNumber} is now in progress.");
            }
            else if (status == ComplaintStatus.Resolved)
            {
                await NotifyCitizenAsync(
                    complaint,
                    NotificationType.ComplaintResolved,
                    $"Your complaint {complaint.RefNumber} has been resolved. Please share feedback.");
            }
        }

        private async Task NotifyMessageRecipientsAsync(Complaint complaint, ComplaintMessage message)
        {
            if (message.SenderType == SenderType.Citizen)
            {
                if (complaint.AssignedAgentId.HasValue)
                {
                    await NotifyUserAsync(
                        SenderType.Agent,
                        complaint.AssignedAgentId.Value,
                        complaint.Id,
                        NotificationType.MessageReceived,
                        $"New citizen message on complaint {complaint.RefNumber}.");
                    return;
                }

                var assigners = await _userRepo.GetByRoleAsync(UserRole.Assigner);
                await NotifyUsersAsync(
                    assigners.Select(u => u.Id),
                    SenderType.Agent,
                    complaint.Id,
                    NotificationType.MessageReceived,
                    $"New citizen message on complaint {complaint.RefNumber}.");
            }
            else
            {
                await NotifyCitizenAsync(
                    complaint,
                    NotificationType.MessageReceived,
                    $"New message on your complaint {complaint.RefNumber}.");
            }
        }

        private async Task NotifyCitizenAsync(
            Complaint complaint,
            NotificationType type,
            string message)
        {
            await NotifyUserAsync(SenderType.Citizen, complaint.CitizenId, complaint.Id, type, message);
        }

        private async Task NotifyUserAsync(
            SenderType userType,
            Guid userId,
            Guid complaintId,
            NotificationType type,
            string message)
        {
            await _notificationRepo.AddAsync(new Core.Entities.Notification
            {
                Id = Guid.NewGuid(),
                UserType = userType,
                UserId = userId,
                ComplaintId = complaintId,
                Type = type,
                Message = message,
                IsRead = false,
                CreatedAt = DateTime.UtcNow
            });
            await _pushNotificationService.SendAsync(userType, userId, "Citizen Connect", message, complaintId);
        }

        private async Task NotifyUsersAsync(
            IEnumerable<Guid> userIds,
            SenderType userType,
            Guid complaintId,
            NotificationType type,
            string message)
        {
            var notifications = userIds.Distinct().Select(userId => new Core.Entities.Notification
            {
                Id = Guid.NewGuid(),
                UserType = userType,
                UserId = userId,
                ComplaintId = complaintId,
                Type = type,
                Message = message,
                IsRead = false,
                CreatedAt = DateTime.UtcNow
            }).ToList();

            if (notifications.Count > 0)
            {
                await _notificationRepo.AddRangeAsync(notifications);
                await _pushNotificationService.SendManyAsync(userType, notifications.Select(x => x.UserId), "Citizen Connect", message, complaintId);
            }
        }
    }
}

File: citizen_connect_python/app/services/complaint_service.py
from datetime import datetime, timezone
from sqlalchemy.ext.asyncio import AsyncSession
from fastapi import UploadFile

from app.models.complaint import (
    Complaint, ComplaintMedia, ComplaintMessage, ComplaintITopMapping
)
from app.repositories.complaint_repository import ComplaintRepository
from app.repositories.internal_user_repository import InternalUserRepository
from app.repositories.location_repository import LocationRepository
from app.schemas.complaint import (
    SubmitComplaintSchema, AssignDepartmentSchema, UpdateStatusSchema,
    ComplaintResponseSchema, ComplaintMediaResponseSchema,
    PagedComplaintsSchema, ComplaintStatus, MediaTypeEnum, SenderTypeEnum
)
from app.services.storage_service import LocalStorageService
from app.services.itop_adapter import ITopTicketAdapter, ITopTicketCreateRequest
import uuid


class ComplaintService:
    def __init__(self, db: AsyncSession):
        self.db = db
        self.complaint_repo = ComplaintRepository(db)
        self.user_repo = InternalUserRepository(db)
        self.location_repo = LocationRepository(db)
        self.storage = LocalStorageService()
        self.itop_adapter = ITopTicketAdapter()

    async def submit_complaint(
        self,
        citizen_id: uuid.UUID,
        dto: SubmitComplaintSchema
    ) -> ComplaintResponseSchema:
        # Validate block exists
        block = await self.location_repo.get_block_by_id(dto.block_id)
        if not block:
            raise KeyError("Block not found.")

        ref_number = await self.complaint_repo.generate_ref_number()
        now = datetime.now(timezone.utc)

        complaint = Complaint(
            id=uuid.uuid4(),
            ref_number=ref_number,
            title=dto.title,
            description=dto.description,
            category_id=dto.category_id,
            block_id=dto.block_id,
            citizen_id=citizen_id,
            priority=dto.priority,
            status=ComplaintStatus.Submitted,
            submitted_at=now,
            created_at=now,
            updated_at=now
        )

        await self.complaint_repo.create(complaint)

        # Reload with all relations
        created = await self.complaint_repo.get_by_id(complaint.id)
        if not created:
            raise Exception("Failed to retrieve created complaint.")
        await self._create_itop_ticket_mapping(created)
        return self._map_to_schema(created)

    async def get_by_id(self, complaint_id: uuid.UUID) -> ComplaintResponseSchema:
        complaint = await self.complaint_repo.get_by_id(complaint_id)
        if not complaint:
            raise KeyError("Complaint not found.")
        return self._map_to_schema(complaint)

    async def get_my_complaints(
        self, citizen_id: uuid.UUID
    ) -> list[ComplaintResponseSchema]:
        complaints = await self.complaint_repo.get_by_citizen_id(citizen_id)
        return [self._map_to_schema(c) for c in complaints]

    async def get_all_complaints(
        self,
        status: int | None,
        department_id: int | None,
        block_id: int | None,
        page: int,
        page_size: int
    ) -> PagedComplaintsSchema:
        items = await self.complaint_repo.get_all(
            status, department_id, block_id, page, page_size
        )
        total = await self.complaint_repo.get_total_count(
            status, department_id, block_id
        )
        total_pages = -(-total // page_size)  # ceiling division
        return PagedComplaintsSchema(
            items=[self._map_to_schema(c) for c in items],
            total_count=total,
            page=page,
            page_size=page_size,
            total_pages=total_pages
        )

    async def assign_department(
        self,
        complaint_id: uuid.UUID,
        dto: AssignDepartmentSchema,
        assigned_by_id: uuid.UUID
    ) -> ComplaintResponseSchema:
        complaint = await self.complaint_repo.get_by_id(complaint_id)
        if not complaint:
            raise KeyError("Complaint not found.")
        if complaint.status != ComplaintStatus.Submitted:
            raise ValueError("Only submitted complaints can be assigned.")

        complaint.assigned_department_id = dto.department_id
        complaint.status = ComplaintStatus.Assigned
        complaint.assigned_at = datetime.now(timezone.utc)
        complaint.updated_at = datetime.now(timezone.utc)

        await self.complaint_repo.update(complaint)
        updated = await self.complaint_repo.get_by_id(complaint_id)
        return self._map_to_schema(updated)

    async def assign_agent(
        self,
        complaint_id: uuid.UUID,
        agent_id: uuid.UUID,
        assigned_by_id: uuid.UUID
    ) -> ComplaintResponseSchema:
        complaint = await self.complaint_repo.get_by_id(complaint_id)
        if not complaint:
            raise KeyError("Complaint not found.")

        agent = await self.user_repo.get_by_id(agent_id)
        if not agent:
            raise KeyError("Agent not found.")

        complaint.assigned_agent_id = agent_id
        if complaint.status == ComplaintStatus.Assigned:
            complaint.status = ComplaintStatus.InProgress
        complaint.updated_at = datetime.now(timezone.utc)

        await self.complaint_repo.update(complaint)
        updated = await self.complaint_repo.get_by_id(complaint_id)
        return self._map_to_schema(updated)

    async def update_status(
        self,
        complaint_id: uuid.UUID,
        dto: UpdateStatusSchema,
        updated_by_id: uuid.UUID
    ) -> ComplaintResponseSchema:
        complaint = await self.complaint_repo.get_by_id(complaint_id)
        if not complaint:
            raise KeyError("Complaint not found.")

        self._validate_status_transition(complaint.status, dto.status)
        complaint.status = dto.status
        complaint.updated_at = datetime.now(timezone.utc)

        if dto.status == ComplaintStatus.Resolved:
            complaint.resolved_at = datetime.now(timezone.utc)
        elif dto.status == ComplaintStatus.Closed:
            complaint.closed_at = datetime.now(timezone.utc)

        await self.complaint_repo.update(complaint)

        # Add system message separately
        message = ComplaintMessage(
            id=uuid.uuid4(),
            complaint_id=complaint_id,
            sender_type=SenderTypeEnum.System,
            sender_id=updated_by_id,
            message=dto.remarks or f"Status updated to {ComplaintStatus.to_string(dto.status)}",
            is_read=False,
            created_at=datetime.now(timezone.utc)
        )
        await self.complaint_repo.add_message(message)

        updated = await self.complaint_repo.get_by_id(complaint_id)
        return self._map_to_schema(updated)

    async def upload_media(
        self,
        complaint_id: uuid.UUID,
        file: UploadFile,
        uploaded_by_id: uuid.UUID
    ) -> ComplaintMediaResponseSchema:
        complaint = await self.complaint_repo.get_by_id(complaint_id)
        if not complaint:
            raise KeyError("Complaint not found.")

        self._validate_upload(file)
        content = await file.read()
        if len(content) == 0:
            raise ValueError("Uploaded file is empty.")
        if len(content) > 50 * 1024 * 1024:
            raise ValueError("Uploaded file exceeds the 50 MB limit.")
        media_type = self.storage.determine_media_type(file.content_type or "")
        folder = f"complaints/{complaint_id}"

        file_path = await self.storage.save_file(
            content,
            file.filename or "upload",
            folder
        )

        media = ComplaintMedia(
            id=uuid.uuid4(),
            complaint_id=complaint_id,
            media_type=media_type,
            file_name=file.filename or "upload",
            file_path=file_path,
            file_size=len(content),
            mime_type=file.content_type,
            created_at=datetime.now(timezone.utc)
        )

        await self.complaint_repo.add_media(media)

        return ComplaintMediaResponseSchema(
            id=media.id,
            media_type=MediaTypeEnum.to_string(media.media_type),
            file_name=media.file_name,
            file_url=self.storage.get_file_url(media.file_path),
            file_size=media.file_size,
            created_at=media.created_at
        )

    # ── Mapper ─────────────────────────────────────────────────────────────

    async def _create_itop_ticket_mapping(self, complaint: Complaint) -> None:
        result = await self.itop_adapter.create_ticket(
            ITopTicketCreateRequest(
                complaint_id=str(complaint.id),
                ref_number=complaint.ref_number,
                title=complaint.title,
                description=complaint.description,
                citizen_name=complaint.citizen.full_name if complaint.citizen else "",
                citizen_phone=complaint.citizen.phone if complaint.citizen else "",
                category_name=complaint.category.name if complaint.category else "",
                block_name=complaint.block.name if complaint.block else "",
                priority=complaint.priority
            )
        )
        if not result.was_attempted:
            return

        mapping = ComplaintITopMapping(
            id=uuid.uuid4(),
            complaint_id=complaint.id,
            itop_ticket_ref=result.ticket_ref or "",
            itop_ticket_id=result.ticket_id or "",
            itop_class="UserRequest",
            last_synced_at=datetime.now(timezone.utc) if result.success else None,
            sync_status=1 if result.success else 2,
            last_sync_error=None if result.success else result.error
        )
        await self.complaint_repo.add_itop_mapping(mapping)

    @staticmethod
    def _validate_status_transition(current: int, new_status: int) -> None:
        allowed = {
            ComplaintStatus.Submitted: {ComplaintStatus.Assigned, ComplaintStatus.Rejected},
            ComplaintStatus.Assigned: {ComplaintStatus.InProgress, ComplaintStatus.Rejected},
            ComplaintStatus.InProgress: {ComplaintStatus.Resolved, ComplaintStatus.Rejected},
            ComplaintStatus.Resolved: {ComplaintStatus.Closed, ComplaintStatus.InProgress},
        }
        if new_status not in allowed.get(current, set()):
            raise ValueError(
                f"Invalid status transition from "
                f"{ComplaintStatus.to_string(current)} to "
                f"{ComplaintStatus.to_string(new_status)}."
            )

    @staticmethod
    def _validate_upload(file: UploadFile) -> None:
        max_bytes = 50 * 1024 * 1024
        if not file.filename:
            raise ValueError("Uploaded file name is required.")
        size = getattr(file, "size", None)
        if size is not None and size <= 0:
            raise ValueError("Uploaded file is empty.")
        if size is not None and size > max_bytes:
            raise ValueError("Uploaded file exceeds the 50 MB limit.")

    def _map_to_schema(self, c: Complaint) -> ComplaintResponseSchema:
        return ComplaintResponseSchema(
            id=c.id,
            ref_number=c.ref_number,
            title=c.title,
            description=c.description,
            status=ComplaintStatus.to_string(c.status),
            priority=c.priority,
            created_at=c.created_at,
            submitted_at=c.submitted_at,
            assigned_at=c.assigned_at,
            resolved_at=c.resolved_at,
            citizen_id=c.citizen_id,
            citizen_name=c.citizen.full_name if c.citizen else "",
            citizen_phone=c.citizen.phone if c.citizen else "",
            block_id=c.block_id,
            block_name=c.block.name if c.block else "",
            category_id=c.category_id,
            category_name=c.category.name if c.category else "",
            assigned_department_id=c.assigned_department_id,
            assigned_department_name=c.assigned_department.name if c.assigned_department else None,
            assigned_agent_id=c.assigned_agent_id,
            assigned_agent_name=c.assigned_agent.full_name if c.assigned_agent else None,
            media=[
                ComplaintMediaResponseSchema(
                    id=m.id,
                    media_type=MediaTypeEnum.to_string(m.media_type),
                    file_name=m.file_name,
                    file_url=self.storage.get_file_url(m.file_path),
                    file_size=m.file_size,
                    created_at=m.created_at
                )
                for m in c.media
            ]
        )
